Kafka Client FindCoordinator implementation

live-wire opened the pull request:
### This pull request contains
- FIndCoordinator implementation that utilizes the `kafka.(*Client)` abstraction.
- A tiny unit test for the same.

---
According to [this design doc](https://docs.google.com/document/d/11Jqy_GjUGtdXJK94XGsEIK7CP1SnQGdp2eF0wSw9ra8/edit#heading=h.dz78fc8xlsd5), the first step in implementing a **Transactional Producer** is looking for a TransactionCoordinator for the specified transactionalId using `FindCoordinatorRequest`. 

---
This PR is a step towards solving #233

CC: @achille-roussel

achille-roussel commented on findcoordinator.go:
```diff
@@ -2,8 +2,120 @@ package kafka
+}
+
+// WaitForCoordinatorIndefinitely is a blocking call till a coordinator is found
+func (c *Client) WaitForCoordinatorIndefinitely(ctx context.Context, req *FindCoordinatorRequest) (*FindCoordinatorResponse, error) {
```
I would prefer to not include this method, and keep the `kafka.(*Client)` API a 1:1 mapping to Kafka APIs.

The one-second blocking call may also not be the right fit for many applications, which could need exponential backoff, or respecting context cancellation, etc...

It isn't very complex for a program to re-implement this logic, maybe we could even put it as an example in the `kafka.(*Client).FindCoordinator` documentation?

live-wire replied:
Makes sense. This shouldn't be a part of the `kafka.(*Client)` API. Will do the needful.

achille-roussel commented on findcoordinator.go:
```diff
@@ -2,8 +2,120 @@ package kafka
+
+// FindCoordinatorResponseCoordinator contains details about the found coordinator
+type FindCoordinatorResponseCoordinator struct {
+	// NodeID holds the broker id.
+	NodeID int32
+
+	// Host of the broker
+	Host string
+
+	// Port on which broker accepts requests
+	Port int32
```
We've taken the approach of not exposing implementation details like the size of the integer types in other public APIs, let's stay consistent?
```suggestion
	// NodeID holds the broker id.
	NodeID int

	// Host of the broker
	Host string

	// Port on which broker accepts requests
	Port int
```

live-wire replied:
Roger that.

achille-roussel commented on findcoordinator.go:
```diff
@@ -2,8 +2,120 @@ package kafka
+	// The error contains both the kafka error code, and an error message
+	// returned by the kafka broker.
+	Error error
+
+	// Error code
+	ErrorCode int
```
Since we already inject the error in the `Error` field, this field seems redundant?
```suggestion
```

live-wire replied:
The current `makeError` implementation returns an `error` object from the errorCode provided.
Once that is done, the error code information can not be accessed again. There should be a way to access error code information as well to ensure downstream functions that make use of `FindCoordinator` have more visibility. 

I will refactor this to still retain error code information.

achille-roussel replied:
I don't think that's accurate, the error returned by `makeError` is either a `kafka.Error`, or wraps one.

The underlying `kafka.Error` can always be retrieved with:
```go
var kafkaError kafka.Error
if errors.As(err, &kafkaErr) {
  // use kafkaError
  ...
}
```

live-wire replied:
Interesting. TIL 😮  
Will make the necessary modification.

achille-roussel reviewed (commented):
Thanks for your contribution @live-wire !

I left a few comments, let me know what you think about it.

live-wire commented:
Hello @achille-roussel ,

I made the following updates:
- Removed WaitForCoordinator function from the `kafka.(*Client)` API.
- Moved the waitingForCoordinator indefinitely to the test file. (The test needs wait for coordinator initialisation).
- Created a new struct to hold error and error code information.
- Made the other suggested fixes.

Please have a look.

achille-roussel commented on error.go:
```diff
@@ -9,6 +9,12 @@ import (
+type ErrorWithCode struct {
+	Error error
+	Code  int16
+}
```
See my response to this discussion https://github.com/segmentio/kafka-go/pull/596#discussion_r576451053

I don't think we need to introduce this new public API.

live-wire replied:
Done.

live-wire commented:
@achille-roussel I made the suggested changes.

achille-roussel commented on findcoordinator.go:
```diff
@@ -2,8 +2,92 @@ package kafka
+		Coordinator: coordinator,
+	}
+
+	return ret, ret.Error
```
The method should only return an error if no responses were available:
```suggestion
	return ret, nil
```

live-wire replied:
Done. 👍

achille-roussel reviewed (approved):
Thanks for the contribution 👍
